### PR TITLE
[Merged by Bors] - feat: port ModelTheory.DirectLimit

### PR DESCRIPTION
**Pull request by GabinKolly**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by QuinnLesquimau** [comment]
A lot of the work was just translating from `Quotient.mk'` to `Quotient.mk` and back, perhaps it would be easier to define a specific map for direct limits of structures.
I did these translations with the tactic `change`, which is probably ugly.

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+namespace DirectLimit
+
+-- Porting note : Instead of `Σi, G i`, we use the alias SigmaG which depends on `f`. This way,
+-- Lean can infer what `L` is in the `setoid` instance. Otherwise we have a
+-- "cannot find synthesization order" error. See the discussion at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/local.20instance.20cannot.20find.20synthesization.20order.20in.20porting
+set_option linter.unusedVariables false
+/-- Alias for `Σi, G i`. -/
+def SigmaG (f : ∀ i j, i ≤ j → G i ↪[L] G j) := Σi, G i
+set_option linter.unusedVariables true
```
Comment (Ruben-VandeVelde) [inline Mathlib/ModelTheory/DirectLimit.lean:91]:
```suggestion
set_option linter.unusedVariables false in
/-- Alias for `Σi, G i`. -/
def SigmaG (f : ∀ i j, i ≤ j → G i ↪[L] G j) := Σi, G i
```

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 2]
```diff
@@ -59,32 +59,41 @@ variable {G' : ℕ → Type w} [∀ i, L.Structure (G' i)] (f' : ∀ n : ℕ, G'
...
 
 namespace DirectLimit
 
+-- Porting note : Instead of `Σi, G i`, we use the alias SigmaG which depends on `f`. This way,
+-- Lean can infer what `L` is in the `setoid` instance. Otherwise we have a
+-- "cannot find synthesization order" error. See the discussion at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/local.20instance.20cannot.20find.20synthesization.20order.20in.20porting
+
+/-- Alias for `Σi, G i`. -/
+@[nolint unusedArguments]
+def SigmaG (_f : ∀ i j, i ≤ j → G i ↪[L] G j) := Σi, G i
```
Comment (jcommelin) [inline Mathlib/ModelTheory/DirectLimit.lean:91]: Just thinking out loud: should this be called something like `SigmaStr`? I'm not a big fan of names that hardcode some piece of notation, `G` is a variable after all.

How about `Language.Structure.Sigma`? Then you can write `G.Sigma`.

Also, it seems that the `Setoid` instance below doesn't just depend on `L`, but in fact depends on `f`. Which is a stronger reason to include `f` in the type. Please mention that in the porting note above. (In fact, please add that explanation to the docstring of this definition.

**Comment by kmill** [comment]
bors d=jcommelin

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**